libkbfs/tlf_edit_history.go

```diff
@@ -348,26 +434,168 @@ func (teh *TlfEditHistory) GetComplete(ctx context.Context,
+	removed := make(map[string]bool)
+	// TODO: can I used chains.deletedOriginals instead?  It's hard to
+	// get the full path that way.
+	for _, chain := range chains.byOriginal {
+		for _, op := range chain.ops {
+			rop, ok := op.(*rmOp)
+			if !ok {
+				continue
+			}
+			path := rop.getFinalPath().String() + "/" + rop.OldName
+			// A rename op might show later that this was only renamed.
+			removed[path] = true
 		}
-		currEdits[w] = list
 	}
-	teh.log.CDebugf(ctx, "Edits complete: %d revisions, starting from "+
-		"revision %d", len(rmds), rmds[0].Revision)
 
-	teh.lock.Lock()
-	defer teh.lock.Unlock()
-	teh.edits = currEdits
-	teh.editsTime = teh.config.Clock().Now()
-	return teh.getEditsCopyLocked(), nil
+	// Which paths have been renamed?
+	renamed := make(map[string]string)
+	for original, ri := range chains.renamedOriginals {
+		// Find both the old path and the new path using the old and
+		// new parents (which are each guaranteed to have at least one
+		// op, since the rename operation is split up into two ops).
+		oldParentChain, ok := chains.byOriginal[ri.originalOldParent]
+		if !ok || len(oldParentChain.ops) == 0 {
+			teh.log.CDebugf(ctx, "Couldn't find old parent to a rename "+
+				"op for original ptr %v", original)
+		}
+		oldPath := oldParentChain.ops[0].getFinalPath().String() + "/" +
+			ri.oldName
+
+		newParentChain, ok := chains.byOriginal[ri.originalNewParent]
+		if !ok || len(newParentChain.ops) == 0 {
+			teh.log.CDebugf(ctx, "Couldn't find new parent to a rename "+
+				"op for original ptr %v", original)
+		}
+		newPath := newParentChain.ops[0].getFinalPath().String() + "/" +
+			ri.newName
+
+		// Ignore any previous rmOps.
+		delete(removed, oldPath)
+		// If a file was overwritten, ignore all the old edits.
+		removed[newPath] = true
+		// Rename the file.
+		renamed[oldPath] = newPath
+	}
+
+	// Also, remove old edits for new file paths, because the newer
+	// edits take precedence.
+	for _, edits := range newEdits {
+		for _, edit := range edits {
+			removed[edit.Filepath] = true
+			delete(renamed, edit.Filepath)
+		}
+	}
+
+	// Remove and rename old edits as needed.
+	if len(removed)+len(renamed) > 0 {
+		teh.log.CDebugf(ctx, "Removed paths: %v, renamed paths: %v",
+			removed, renamed)
+		currEdits.updateOldEdits(removed, renamed)
+	}
+
+	currEdits.addNewEdits(newEdits)
+	// If we have a net negative removed, we have to search back
+	// farther in time to become complete again.
+	if !currEdits.isComplete() && wasComplete {
+		teh.log.CDebugf(ctx, "Too many removals; re-calculating edit history")
+		func() {
+			teh.lock.Lock()
+			defer teh.lock.Unlock()
+			teh.edits = nil
+		}()
+		_, err := teh.GetComplete(ctx, rmds[len(rmds)-1])
+		return err
+	}
+
+	teh.setEdits(ctx, currEdits, rmds)
+	return nil
+}
+
+func (teh *TlfEditHistory) process() {
+	var ctx context.Context
+	func() {
+		teh.lock.Lock()
+		defer teh.lock.Unlock()
+		ctx, teh.cancel = context.WithCancel(context.Background())
+	}()
```

Comment:
I've noticed this pattern a few times for code clarity. Is this kind of closure cost-free? Would it be better to lock, do something, then unlock in the method scope itself?